YunoHost/moulinette
# [enh] Support subcategories

### Current status

Ready for review and feedback

:warning: This PR is built on top of [`bit_of_refactoring`](https://github.com/YunoHost/moulinette/compare/unstable...bit_of_refactoring)

### Problem

(Related to https://dev.yunohost.org/issues/609)
There is a reccuring need for 'subcategories' support in moulinette in CLI. For instance, when implementing certificate management, we would have been happy to be able to have `yunohost domain certificate {status, install, renew}` (with options specific to each action) instead of `yunohost domain {cert-status, cert-install, cert-renew}`.

We now run into the same issue for the migration framework which could have its place under `yunohost tools migrations` but we need one more level for the actions.

### Solution

Tweak the moulinette to have an additional and optional "subcategory" level. Had to use some dirty hacks to have separate sections in the --help (see 'actions' and 'subcategories' in the screenshot below)

![](https://framapic.org/tk0WGUkYYa24/Q6aoU4CUFxzz.png) 

### How to test

Pull this branch, add [this](https://paste.yunohost.org/raw/iyevokacij) at the end of Yunohost's action map, and add a [`trololo.py`](https://paste.yunohost.org/raw/igobikohap) in `yunohost/src/yunohost/`.

Test the following (with and without --help) : 
```
yunohost trololo
yunohost trololo nyaaah
yunohost trololo yolo
yunohost trololo swag
yunohost trololo yolo do
yunohost trololo swag add 42
```
You should also test the api : restart yunohost-api, log in the admin interface, then go to `domain.tld/yunohost/api/trololo/yolo/do`
